flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -155,11 +155,53 @@ public void testTransitionToFinishedOnFailedExecutionGraph() throws Exception {
         }
     }
 
+    @Test
+    public void testEnsureOnlyGlobalTerminalStateCallbacksNoCall() throws Exception {
+        // modify MockExecutingContext to disable the "runIfState()" check
+        MockExecutingContext ctx =
+                new MockExecutingContext() {
+                    @Override
+                    public void runIfState(State expectedState, Runnable action) {
+                        action.run();
+                    }
+                };
+        MockedExecuting exec = new ExecutingStateBuilder().build(ctx);
+
+        ctx.setExpectFinished(
+                archivedExecutionGraph ->
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED)));
+        exec.onEnter();
+        exec.suspend(
+                new RuntimeException("suspend")); // trigger transition to non-global terminal state
+        ctx.close(); // manually close context to trigger callbacks
+
+        assertThat(exec.isOnGloballyTerminalStateCalled(), is(false));
+    }
+
+    @Test
+    public void testEnsureOnlyGlobalTerminalStateCallbacksWithCall() throws Exception {
+        // modify MockExecutingContext to disable the "runIfState()" check
+        MockExecutingContext ctx =
+                new MockExecutingContext() {
+                    @Override
+                    public void runIfState(State expectedState, Runnable action) {
+                        action.run();
+                    }
+                };
+        MockedExecuting exec = new ExecutingStateBuilder().build(ctx);
+        ctx.setExpectFinished(
+                archivedExecutionGraph ->
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
+        exec.getExecutionGraph()
+                .failJob(new RuntimeException()); // trigger transition to global terminal state
+        ctx.close(); // manually close context to trigger callbacks
+        assertThat(exec.isOnGloballyTerminalStateCalled(), is(true));
+    }
```

Comment:
Technically speaking, these tests belong rather to a `StateWithExecutionGraphTest` which tests the general behaviour of the `StateWithExecutionGraph`.